lib/ssz_ex.ex

```diff
@@ -184,6 +192,124 @@ defmodule LambdaEthereumConsensus.SszEx do
+           |> encode_schemas do
+      (fixed_parts ++ variable_parts)
+      |> Enum.join()
+      |> then(&{:ok, &1})
+    end
+  end
+
+  defp analyze_schemas(container, schemas) do
+    schemas
+    |> Enum.reduce({[], 0, []}, fn {key, schema},
+                                   {acc_fixed_size_values, acc_fixed_length, acc_variable_values} ->
+      value = Map.fetch!(container, key)
+
+      if variable_size?(schema) do
+        {[:offset | acc_fixed_size_values], @bytes_per_length_offset + acc_fixed_length,
+         [{value, schema} | acc_variable_values]}
+      else
+        {[{value, schema} | acc_fixed_size_values], acc_fixed_length + get_fixed_size(schema),
+         acc_variable_values}
+      end
+    end)
+  end
+
+  defp encode_schemas(tuple_values) do
+    Enum.map(tuple_values, fn {value, schema} -> encode(value, schema) end)
+    |> flatten_results()
+  end
+
+  defp calculate_offsets(variable_parts, fixed_length) do
+    {offsets, _} =
+      Enum.reduce(variable_parts, {[], fixed_length}, fn element, {res, acc} ->
+        {[{acc, {:int, 32}} | res], byte_size(element) + acc}
+      end)
+
+    offsets
+  end
+
+  defp replace_offsets(fixed_size_values, offsets) do
+    {fixed_size_values, _} =
+      Enum.reduce(fixed_size_values, {[], offsets}, &replace_offset/2)
+
+    fixed_size_values
+  end
+
+  defp replace_offset(:offset, {acc_fixed_list, [offset | rest_offsets]}),
+    do: {[offset | acc_fixed_list], rest_offsets}
+
+  defp replace_offset(element, {acc_fixed_list, acc_offsets_list}),
+    do: {[element | acc_fixed_list], acc_offsets_list}
+
+  defp decode_container(binary, module) do
+    schemas = module.schema()
+    fixed_length = get_fixed_length(schemas)
+    <<fixed_binary::binary-size(fixed_length), variable_binary::bitstring>> = binary
+
+    with {:ok, fixed_parts, offsets} <- decode_fixed_section(fixed_binary, schemas, fixed_length),
+         {:ok, variable_parts} <- decode_variable_section(variable_binary, offsets) do
+      {:ok, struct!(module, fixed_parts ++ variable_parts)}
+    end
+  end
+
+  defp decode_variable_section(binary, offsets) do
+    offsets
+    |> Enum.chunk_every(2, 1)
+    |> Enum.reduce({binary, []}, fn
+      [{offset, {key, schema}}, {next_offset, _}], {rest_bytes, acc_variable_parts} ->
+        size = next_offset - offset
+        <<chunk::binary-size(size), rest::bitstring>> = rest_bytes
+        {rest, [{key, decode(chunk, schema)} | acc_variable_parts]}
+
+      [{_offset, {key, schema}}], {rest_bytes, acc_variable_parts} ->
+        {<<>>, [{key, decode(rest_bytes, schema)} | acc_variable_parts]}
+    end)
+    |> then(fn {<<>>, variable_parts} ->
+      flatten_container_results(variable_parts)
+    end)
+  end
+
+  defp decode_fixed_section(binary, schemas, fixed_length) do
+    schemas
+    |> Enum.reduce({binary, [], []}, fn {key, schema},
+                                        {rest_bytes, acc_fixed_parts, acc_offsets} ->
```

Comment:
Nit: Let's change the names of the accumulator members to binary, fixed_parts and offsets. It's understandable that we're in an iterating function and they are accumulators, and also the `rest` in `rest_bytes` makes it a bit confusing with the `rest` variable we create later.